src/clang-c-frontend/expr2ccode.cpp

```diff
@@ -0,0 +1,1509 @@
+    dest += ";\n";
+  }
+  dest += "}";
+
+  if(is_typedef_struct_union(tag))
+    dest += " " + tag;
+
+  return dest;
+}
+
+std::string expr2ccodet::convert_rec(
+  const typet &src,
+  const c_qualifierst &qualifiers,
+  const std::string &declarator)
+{
+  c_qualifierst new_qualifiers(qualifiers);
+  new_qualifiers.read(src);
+
+  // Constructing qualifiers that go before the type
+  std::string q = "";
+  if(new_qualifiers.is_constant)
+    q += "const ";
+  if(new_qualifiers.is_volatile)
+    q += "volatile ";
+
+  // The "__restrict" qualifier is used with pointers and goes
+  // before the pointer name
+  std::string rst_q = "";
+  if(new_qualifiers.is_restricted)
+    rst_q += "__restrict ";
+
+  std::string d = declarator == "" ? declarator : " " + declarator;
+
+  if(src.is_bool())
+  {
+    return q + "_Bool" + d;
+  }
+  if(src.id() == "empty")
+  {
+    return q + "void" + d;
+  }
+  else if(src.id() == "signedbv" || src.id() == "unsignedbv")
+  {
+    BigInt width = string2integer(src.width().as_string());
+
+    // Sometimes there is no way of distinguishing between
+    // types based on their width. For example,
+    // "long" and "long long" are both 64 bits on
+    // 64-bit platforms. However, in many cases the type contains
+    // a non-empty "#cpp_type" field with a string
+    // representation of the corresponding type in the original
+    // program. Hence, if "#cpp_type" is not empty, we try to use
+    // this information first, and only resort to infering the type name
+    // from its width otherwise
+    std::string cpp_type = src.get("#cpp_type").as_string();
+    if(!cpp_type.empty() && width % 8 == 0)
+    {
+      std::replace(cpp_type.begin(), cpp_type.end(), '_', ' ');
+      size_t sign_pos = cpp_type.find(" ");
+      std::string sign_str = cpp_type.substr(0, sign_pos + 1);
+      if(sign_str == "signed ")
+        sign_str = "";
+      std::string type_name = cpp_type.substr(sign_pos + 1, cpp_type.length());
+      return q + sign_str + type_name + d;
+    }
+
+    // We cannot have a signed type with a 1 bit width.
+    // So setting "sign_str" to "unsigned" to allow _ExtInt(1)
+    bool is_signed = src.id() == "signedbv";
+
+    // Also we do not add the "signed" keyword for signed types.
+    // Only unsigned is being added.
+    std::string sign_str = (is_signed && width > 1) ? "" : "unsigned ";
+
+    if(width == config.ansi_c.int_width)
+      return q + sign_str + "int" + d;
+
+    if(width == config.ansi_c.long_int_width)
+      return q + sign_str + "long int" + d;
+
+    if(width == config.ansi_c.long_long_int_width)
+      return q + sign_str + "long long int" + d;
+
+    if(width == config.ansi_c.char_width)
+      return q + sign_str + "char" + d;
+
+    if(width == config.ansi_c.short_int_width)
+      return q + sign_str + "short int" + d;
+
+    return q + sign_str + "_ExtInt(" + std::to_string(width.to_uint64()) + ")" +
```

Comment:
`_ExtInt` is a clang-only feature to my knowledge aimed at integer types with more than 64 bits. Would it be possible to include GCC here, as well for the most common situation this case is triggered: bitfields? I realize, this would require knowledge of whether or not this declaration is part of a struct/union definition. So if it's too much work to support for now, I would again suggest to add a comment about that here. Also, `_ExtInt` is actually [deprecated](https://reviews.llvm.org/D108643), though for compatibility reasons it might still be the best option for goto2c. This reasoning should be mentioned as well.

Reply:
Updated as suggested.